src/webgpu/api/validation/capability_checks/limits/maxBindGroupsPlusVertexBuffers.spec.ts

```diff
@@ -0,0 +1,274 @@
+  device: GPUDevice,
+  preference: VertexBufferBindGroupPreference,
+  testValue: number
+) {
+  switch (preference) {
+    case 'bindGroups': {
+      const numBindGroups = Math.min(testValue, device.limits.maxBindGroups);
+      const numVertexBuffers = Math.max(0, testValue - numBindGroups);
+      return { numVertexBuffers, numBindGroups };
+    }
+    case 'vertexBuffers': {
+      const numVertexBuffers = Math.min(testValue, device.limits.maxVertexBuffers);
+      const numBindGroups = Math.max(0, testValue - numVertexBuffers);
+      return { numVertexBuffers, numBindGroups };
+    }
+  }
+}
+
+function createLayout(device: GPUDevice, layoutType: LayoutType, numBindGroups: number) {
+  switch (layoutType) {
+    case 'auto':
+      return 'auto';
+    case 'explicit': {
+      const bindGroupLayouts = new Array(numBindGroups);
+      if (numBindGroups > 0) {
+        bindGroupLayouts.fill(device.createBindGroupLayout({ entries: [] }));
+        bindGroupLayouts[numBindGroups - 1] = device.createBindGroupLayout({
+          entries: [
+            {
+              binding: 0,
+              visibility: GPUShaderStage.VERTEX,
+              buffer: {},
+            }
+          ],
+        });
+      }
+      return device.createPipelineLayout({ bindGroupLayouts })
+    }
+  }
+}
+
+/**
+ * Generate a render pipeline that can be used to test maxBindGroupsPlusVertexBuffers
+ */
+function getPipelineDescriptor(device: GPUDevice, preference: VertexBufferBindGroupPreference, testValue: number, layoutType: LayoutType) {
+  // Get the numVertexBuffers and numBindGroups we could use given testValue as a total.
+  // We will only use 1 of each but we'll use the last index.
+  const { numVertexBuffers, numBindGroups } = getNumBindGroupsAndNumVertexBuffers(
+    device,
+    preference,
+    testValue
+  );
+
+  const layout = createLayout(device, layoutType, numBindGroups);
+
+  const [bindGroupDecl, bindGroupUsage] =
+    numBindGroups === 0
+      ? ['', '']
+      : [`@group(${numBindGroups - 1}) @binding(0) var<uniform> u: f32;`, `_ = u;`];
+
+  const [attribDecl, attribUsage] =
+    numVertexBuffers === 0
+      ? ['', '']
+      : ['@location(0) v: vec4f', `_ = v; // will use vertex buffer ${numVertexBuffers - 1}`];
+
+  const code = `
+    ${bindGroupDecl}
+
+    @vertex fn vs(${attribDecl}) -> @builtin(position) vec4f {
+      ${bindGroupUsage}
+      ${attribUsage}
+      return vec4f(0);
+    }
+  `;
+
+  const module = device.createShaderModule({ code });
+  const buffers = new Array(numVertexBuffers);
+  if (numVertexBuffers > 0) {
+    buffers[numVertexBuffers - 1] = {
+      arrayStride: 16,
+      attributes: [{ shaderLocation: 0, offset: 0, format: 'float32' }],
+    };
+  }
+
+  return {
+    code,
+    descriptor: {
+      layout,
+      vertex: {
+        module,
+        buffers,
+      },
+      fragment: {
+        module,
+        targets: [{ format: 'rgba8unorm' }],
+      },
+    } as GPURenderPipelineDescriptor,
```

Comment:
nit: avoid unsafe cast
```suggestion
    } as const,
```